### PR TITLE
Extract an interface for a ConnectionUser

### PR DESCRIPTION
**Pull request by swankjesse**
This is working towards supporting pre-warmed connections in ConnectionPool. We want to create connections without a Call object.

### CODE REVIEW

File: okhttp/src/main/kotlin/okhttp3/internal/connection/RealConnectionPool.kt [hunk 1]
```diff
@@ -71,7 +71,7 @@ class RealConnectionPool(
   }
 
   /**
-   * Attempts to acquire a recycled connection to [address] for [call]. Returns the connection if it
+   * Attempts to acquire a recycled connection to [address] for [connectionUser]. Returns the connection if it
    * was acquired, or null if no connection was acquired. The acquired connection will also be
    * assigned to [RealCall.connection].
```
Comment (yschimke) [inline okhttp/src/main/kotlin/okhttp3/internal/connection/RealConnectionPool.kt:76]: nit: doesn't always hold. This is through `connectionUser.acquireConnectionNoEvents()` which may no longer be a call.

**Comment by yschimke** [comment]
I can see where it's going. But it's hard to reason about the interface without the new usage.

**Comment by ean5533** [comment]
@yschimke Here's an incomplete example of what we're trying to accomplish, which is enabled by this refactor: https://github.com/square/okhttp/commit/63a69885808f8ff75e7ce351e4dd1e6df1854ec5

File: okhttp/src/main/kotlin/okhttp3/internal/connection/CallConnectionUser.kt [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+import okhttp3.Protocol
+import okhttp3.Route
+import okhttp3.internal.http.RealInterceptorChain
+
+/**
+ * A connection user that is a specific [RealCall].
+ */
+internal class CallConnectionUser(
+  private val call: RealCall,
+  private val poolConnectionListener: ConnectionListener,
```
Comment (yschimke) [inline okhttp/src/main/kotlin/okhttp3/internal/connection/CallConnectionUser.kt:21]: Should the concrete instances on this interface be responsible for notifying the connection pool?

Seems like logic should match regardless of the impl.

Maybe pull out or wrap with a poolconnectionuser?

#tioli

**Pull request merged**